src/Microsoft.Extensions.Logging.Abstractions/Internal/TypeNameHelper.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace Microsoft.Extensions.Logging.Abstractions.Internal
+{
+    public class TypeNameHelper
+    {
+        private static readonly Dictionary<Type, string> _builtInTypeNames = new Dictionary<Type, string>
+            {
+            { typeof(bool), "bool" },
+            { typeof(byte), "byte" },
+            { typeof(char), "char" },
+            { typeof(decimal), "decimal" },
+            { typeof(double), "double" },
+            { typeof(float), "float" },
+            { typeof(int), "int" },
+            { typeof(long), "long" },
+            { typeof(object), "object" },
+            { typeof(sbyte), "sbyte" },
+            { typeof(short), "short" },
+            { typeof(string), "string" },
+            { typeof(uint), "uint" },
+            { typeof(ulong), "ulong" },
+            { typeof(ushort), "ushort" }
+            };
+
+        public static string GetTypeDisplayName(Type type)
+        {
+            if (type.GetTypeInfo().IsGenericType)
+            {
+                var fullName = type.GetGenericTypeDefinition().FullName;
+
+                // Nested types (public or private) have a '+' in their full name
+                var parts = fullName.Split('+');
+
+                // Handle nested generic types
+                // Examples:
+                // ConsoleApp.Program+Foo`1+Bar
+                // ConsoleApp.Program+Foo`1+Bar`1
+                var partNames = new List<string>(parts.Length);
```

Comment:
Instead of a new list, could you re-use the same array and replace anything that needs to be replaced? (And leave everything else alone.)